Add blog for image gallery (#187)
xataio/mdx-blog: build-image-gallery-astro-cloudflare.mdx

		// Redirect user to the next step
  } catch (e) {
    // pass
  }
}
---

<form method="post" autocomplete="off">
  <Upload selector="user" />
  <input required name="name" type="text" placeholder="Name" />
  <input
    required
    name="slug"
    type="text"
    placeholder="Slug (e.g. rishi-raj-jain)"
  />
  <button type="submit">
    Create Profile &rarr;
  </button>
</form>
```

## Handling image uploads server-side with the Xata SDK

![(/photograph/create - where users upload their photographs)](/images/build-image-gallery/image13.png)

As Cloudflare Pages request body size allows up to 100 MB, you’ll be able to handle image uploads on the server-side. Create an Astro endpoint (`src/pages/api/upload/index.ts`) to receive POST requests containing image binaries and use the Xata SDK to store them in the `uploads` table.

After doing sanity checks on the request body, first create a new (empty) record in your `uploads` table, and then use it as a reference to place the image (buffer) using the Xata TypeScript SDK. Once successfully completed, the endpoint responds with the image’s `public URL`, `height` and `width` back to the front-end to include in the form fields.

> ```tsx

Author: I think this one is going in the `index.ts` file, so I thought it was Typescript. Should I change it to Astro?

cd image-gallery-cms-with-astro-xata-cloudflare
pnpm install
```

## Configure Xata with Astro

To use Xata with Astro seamlessly, install the Xata CLI globally:

```bash
npm install @xata.io/cli -g
```

Then, authorize the Xata CLI so it is associated with the logged in account:

```bash
xata auth login
```

![Authorize the CLI](/images/build-image-gallery/image11.png)

Great! Now, initialize your project locally with the Xata CLI command:

```bash
xata init --db https://Rishi-Raj-Jain-s-workspace-80514q.ap-southeast-2.xata.sh/db/image-gallery-cms-with-xata-astro-cloudflare
```

Answer some quick one-time questions from the CLI to integrate with Astro.

![Xata init with Astro project](/images/build-image-gallery/image12.png)

## Implementing form actions in Astro

You can also allow transitions on form submissions by adding the `ViewTransitions` component.
Here’s an example of the enabled form actions with view transitions in `src/layouts/Layout.astro`:

```tsx
---
// File: src/layouts/Layout.astro

import { ViewTransitions } from "astro:transitions";
---

<html>
  <head>
    <ViewTransitions />
    <!-- stuff here -->
  </head>
  <body>
    <!-- stuff here -->
  </body>
</html>

> ```

Reviewer: All of these references are for `astro` syntax, not `tsx` syntax.

    response.form = true

    // Get the user email from the form submissions
    const data = await Astro.request.formData()

		// Get the user slug, name, and image: URL, width, and height from the form submissions
    const userSlug = data.get('slug') as string
    const userName = data.get('name') as string
    const userImage = data.get('custom_upload_user__uploaded_image_url') as string
    const userImageW = data.get('custom_upload_user__uploaded_w') as string
    const userImageH = data.get('custom_upload_user__uploaded_h') as string

		// Create a blur url of the user image

		// Create the user record with the slug

		// Redirect user to the next step
  } catch (e) {
    // pass
  }
}
---

<form method="post" autocomplete="off">
  <Upload selector="user" />
  <input required name="name" type="text" placeholder="Name" />
  <input
    required
    name="slug"
    type="text"
    placeholder="Slug (e.g. rishi-raj-jain)"
  />
  <button type="submit">
    Create Profile &rarr;
  </button>
</form>
```

## Handling image uploads server-side with the Xata SDK

![(/photograph/create - where users upload their photographs)](/images/build-image-gallery/image13.png)

> (/photograph/create - where users upload their photographs)

Reviewer 2: ```suggestion
```

Reviewer 2: This is duplicate and should be removed

														// Filter the results to the specific profile
														.filter({ 'profile-slug': profile })
														// Get all the photographs
														.getAll()
---

<Layout className="flex flex-col">
  <div class="columns-1 gap-0 md:columns-2 lg:columns-3">
    {
      profilePhotographs.map(
        ({ width: photoW, height: photoH, name: photoName, image: photoImageURL, tagline: photoTagline, imageHash: photoImageHash }, _) => (
					{/* Destructure the width and height to prevent CLS */}
          <img
            width={photoW}
            height={photoH}
            alt={photoName}
            src={photoImageURL}
						{/* Do not lazy load the first image that's loaded into the DOM */}
            loading={_ === 0 ? 'eager' : 'lazy'}
            class="transform bg-cover bg-center bg-no-repeat will-change-auto"
						{/* Create a blur effect with the imageHash stored */}
            style={`background-image: url(${photoImageHash}); transform: translate3d(0px, 0px, 0px);`}
          />
        ),
      )
    }
  </div>
</Layout>
```

![(Image gallery - presenting blur images initially)](/images/build-image-gallery/image15.png)

> (Image gallery - presenting blur images initially)

Reviewer 2: ```suggestion
```

> ![(Image gallery - after image loads complete)](/images/build-image-gallery/image15.png)

Reviewer 2: This is the same image as the one on line 414

So:
Image gallery - presenting blur images initially
Image gallery - after image loads complete

are the same. What did we mean to show here?

> (Image gallery - after image loads completely)

Reviewer 2: ```suggestion
```

## More information

For a more detailed insights, explore the references cited in this post.

| Demo Image Gallery          | https://image-gallery-cms-with-astro-xata-cloudflare.pages.dev/rishi           |
| --------------------------- | ------------------------------------------------------------------------------ |
| GitHub Repo                 | https://github.com/rishi-raj-jain/image-gallery-cms-with-astro-xata-cloudflare |
| Astro with Xata             | https://xata.io/docs/getting-started/astro                                     |
| Astro View Transition Forms | https://docs.astro.build/en/guides/view-transitions/#transitions-with-forms    |
| Xata File Attachments       | https://xata.io/docs/sdk/file-attachments#upload-a-file-using-file-apis        |
| Xata Transformations        | https://xata.io/docs/sdk/image-transformations                                 |
| Xata Get Records            | https://xata.io/docs/sdk/get#the-typescript-sdk-functions-for-querying         |
| Cloudflare Workers Limits   | https://developers.cloudflare.com/workers/platform/limits/#request-limits      |

## What's next?

> We'd love to hear from you if you have any feedback on this tutorial, would like to know more about Xata, or if you'd like to contribute a community blog or tutorial. Reach out to us on [Discord](https://discord.com/invite/kvAcQKh7vm) or join us on [X | Twitter](https://twitter.com/xata). Happy building 🦋

Reviewer 2: ```suggestion
We'd love to hear from you if you have any feedback on this tutorial, would like to know more about Xata, or if you'd like to contribute a community blog or tutorial. Reach out to us on [Discord](https://xata.io/discord) or join us on [X | Twitter](https://twitter.com/xata). Happy building 🦋
```
